Framework/Geometry/src/Objects/MeshObject.cpp

```diff
@@ -0,0 +1,725 @@
+#include "MantidGeometry/Rendering/vtkGeometryCacheWriter.h"
+#include "MantidKernel/Exception.h"
+#include "MantidKernel/make_unique.h"
+#include "MantidKernel/Material.h"
+#include "MantidKernel/PseudoRandomNumberGenerator.h"
+
+#include <boost/make_shared.hpp>
+
+namespace Mantid {
+namespace Geometry {
+
+using Kernel::Material;
+using Kernel::V3D;
+using Kernel::Quat;
+
+/**
+*  Default constuctor
+*/
+MeshObject::MeshObject() : MeshObject("") {}
+
+MeshObject::MeshObject(const std::string &string)
+    : m_boundingBox(), m_object_number(0), m_handler(), m_string(string),
+      m_id(), m_material(), m_triangles(), m_vertices() {}
+
+/**
+* Copy constructor
+* @param A :: The object to initialise this copy from
+*/
+MeshObject::MeshObject(const MeshObject &A) : MeshObject() { *this = A; }
+
+/**
+* Assignment operator
+* @param A :: Object to copy
+* @return *this
+*/
+MeshObject &MeshObject::operator=(const MeshObject &A) {
+  if (this != &A) {
+    m_boundingBox = A.m_boundingBox;
+    m_object_number = A.m_object_number;
+    m_string = A.m_string;
+    m_handler = A.m_handler->clone();
+    m_id = A.m_id;
+    m_material = Kernel::make_unique<Material>(A.material());
+    m_triangles = A.m_triangles;
+    m_vertices = A.m_vertices;
+  }
+  return *this;
+}
+
+MeshObject::~MeshObject() = default;
+
+/*
+ * Initialise with vectors.
+ * @param faces :: Triangles specified by vertex index in anticlockwise order
+ * as seen from outside.
+ * @param vertices :: Vertex positions
+ */
+void MeshObject::initialize(const std::vector<int> &faces,
+                            const std::vector<V3D> &vertices) {
+  if (m_vertices.size() == 0) {
+    m_vertices = vertices;
+    m_triangles = faces;
+    m_handler = boost::make_shared<CacheGeometryHandler>(this);
+  } else {
+    throw std::runtime_error("MeshObject already initialized");
+  }
+}
+
+/**
+ * @param material The new Material that the object is composed from
+ */
+void MeshObject::setMaterial(const Kernel::Material &material) {
+  m_material = Mantid::Kernel::make_unique<Material>(material);
+}
+
+/**
+ * @return The Material that the object is composed from
+ */
+const Kernel::Material MeshObject::material() const {
+  if (m_material)
+    return *m_material;
+  else
+    return Material();
+}
+
+/**
+* Returns whether this object has a valid shape
+* @returns True if the entire MeshObject may enclose
+* one or more volumes.
+*/
+bool MeshObject::hasValidShape() const {
+  // May enclose volume if there are at
+  // at least 4 triangles and 4 vertices
+  return (numberOfTriangles() >= 4 && numberOfVertices() >= 4);
+}
+
+/**
+* Determines whether Pt is within the object or on the surface
+* @param Pt :: Point to be tested
+* @returns 1 if true and 0 if false
+*/
+bool MeshObject::isValid(const Kernel::V3D &Pt) const {
+
+  BoundingBox bb = getBoundingBox();
+  if (!bb.isPointInside(Pt)) {
+    return false;
+  }
+
+  V3D direction(0.0, 0.0, 1.0); // direction to look for intersections
```

Comment:
I'm confused by this, method. We only check along the z-axis for intersections with the triangles of the mesh?

Reply:
The purpose of this code is to check whether a point is inside. We need only check along one direction for this. I chose upwards for this direction.

Reply:
Ok. so the assumption is that the mesh always forms a closed surface.

Reply:
Yes

Reply:
Thanks for clarifying @KarlPalmen. I think that this assumption needs to be clearly stated, perhaps in the documentation on the header file.